### ISSUE TITLE
Make attachments immutable

### ISSUE THREAD

**Issue by chetanmeh** [apache/incubator-openwhisk#3500]
As part of work to enable AttachmentStore SPI(#3450) we need to make attachments immutable. More details are provided in [mail thread][1]. In brief key point is that each attachment must be give a unique name and it should be created in following way

1. Upload the attachment and have ArtifactStore return a generated name
2. Then update the document with attachmentName set to name returned in previous step
3. Then delete the old attachment after #2 completes successfully

The approach would allow an AttachmentStore to

1. Ensure that attachments are uniquely named
2. Possibly inline small attachments by encoding the attachment content in attachment name say if 1KB doc is to be inlined the attachment name is hex encoded doc content

[1]: https://lists.apache.org/thread.html/f90129cf991023c5a2c4622658d37d0eb14ecbc561f970e55bbd0035@%3Cdev.openwhisk.apache.org%3E

**Comment by chetanmeh** [comment]
I tried to implement this but the way CouchDB handled attachment I am not able to find a generic approach to implement the above flow. 

With CouchDB following problems are seen

1. Attaching attachment requires the document to be present - Its not possible to upload an attachment and then link to the document. So for initial creation we need to first create the document
2. Updating document post attach causes the attachment to be lost - If a document with attachment is updated later then attachment gets lost. Looks like one needs to [pass in a stub][1] of attachment for such an update to work
    > When you want to update the document (e.g. through PUT) you must include the "stub", otherwise Couch will remove the attachment.

Given these limitation it looks like we cannot implement proposed logic in a generic way. For now I see following option

#### A - Store specific flow in DocumentFactory

Use current approach when just `CouchDBRestStore` is configured and use the proposed approch when any other `AttachmentStore` is configured

#### B - Introduce a new method in ArtifactStore

Another option would be to be add another method in ArtifactStore

```scala
  protected[database] def putWithAttachment(d: DocumentAbstraction,
                              attachedUpdater: (Attached, DocumentAbstraction) => DocumentAbstraction,
                              contentType: ContentType,
                              docStream: Source[ByteString, _],
                              oldAttached: Option[Attached])(implicit transid: TransactionId): Future[DocInfo]
```

where
* `attachedUpdater` - Method which would update the `DocumentAbstraction` with new attachment details
* `oldAttached` - Details about old attachment. We can also just accept an attachmentName string also here

Such a method can then be efficiently implemented by respective `ArtifactStore`  for e.g. with
* `CouchDBRestStore`
    1. Generate a random name
    2. Update the DocumentAbstraction and save it
    3. Now attach the passed source with generated name
    4. No need for deleting old attachment as post document update CouchDB would not remember it anyway
* `CouchDBRestStore` + `S3AttachmentStore`
    1. Generate a random name and then upload to S3 with generated name
    2. Update the DocumentAbstraction and save it
    3. Delete the old attachment

@rabbah @markusthoemmes Any thoughts on which approach to take
[1]: https://stackoverflow.com/a/34325095/1035417

**Comment by chetanmeh** [comment]
Resolving as Closed given #3502 is merged. OpenWhisk now does not modify an attachment and instead add a new and remove older attachment, So each attachment gets a unique name upon any change
